tools/eth-migration/fetch-offchain-index-keys.mjs

```diff
@@ -0,0 +1,85 @@
+// Only way to silence PolkadotJS API warnings we don't want
+console.warn = () => {};
+
+import { ApiPromise, WsProvider } from "@polkadot/api";
+
+const SOURCE_URL = "wss://1.rpc.frequency.xyz";
+const PROMISE_BATCH_SIZE = 500;
+
+const options = {
+	rpc: {
+		msa: {
+			getKeysByMsaId: {
+				description: 'Fetch Keys for an MSA Id',
+				params: [
+					{
+						name: 'msa_id',
+						type: 'MessageSourceId',
+					},
+				],
+				type: 'Option<KeyInfoResponse>',
+			},
+		},
+	},
+	types: {
+		MessageSourceId: 'u64',
+		KeyInfoResponse: {
+			msa_keys: 'Vec<AccountId>',
+			msa_id: 'MessageSourceId',
+		},
+	},
+};
+
+export async function fetchControlKeysFromOffchainRpc(sourceUrl) {
+	// Connect to the state source
+	const sourceProvider = new WsProvider(sourceUrl);
+	const sourceApi = await ApiPromise.create({provider: sourceProvider, ...options});
+
+	// Get the max schema id
+	let msaMaxId = (await sourceApi.query.msa.currentMsaIdentifierMaximum()).toNumber();
+	let promises = [];
+	let id = 1;
+	let batchNumber = 0;
+	for (; id <= msaMaxId ; id++){
+		promises.push(sourceApi.rpc.msa.getKeysByMsaId(id));
+		if (id % PROMISE_BATCH_SIZE === 0 && promises.length > 0) {
+			printRpcResults(await Promise.all(promises), batchNumber);
+			promises = [];
+			msaMaxId = (await sourceApi.query.msa.currentMsaIdentifierMaximum()).toNumber();
+			batchNumber++;
+		}
+	}
+
+	if (promises.length > 0) {
+		printRpcResults(await Promise.all(promises), batchNumber);
+	}
+}
+
+function printRpcResults(results, batchNumber) {
+	let index = 0;
+	for (const r of results) {
+		if (!r.isSome) {
+			console.error(`No keys for MsaId: ${(batchNumber * PROMISE_BATCH_SIZE + index)}!`);
```

Comment:
I'm not sure how you intend to handle the error case when you're using the sorted output, but I wonder if outputting a valid row with 0x0 as the key, instead of outputting an error might be more helpful.  That or you could redirect errors to another file.

Reply:
So by default you would see the errors in console but you can pipe it into another file if you want. I think it's a matter of taste how you would like to see it . Currently it shows up in the diff and it's not a big deal imo.